src/ddscxx/tests/SharedMemory.cpp

```diff
@@ -0,0 +1,398 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <gtest/gtest.h>
+
+#include "dds/dds.hpp"
+#include "dds/ddsrt/environ.h"
+#include "iceoryx_posh/testing/roudi_gtest.hpp"
+#include "HelloWorldData.hpp"
+#include "Space.hpp"
+#include "Serialization.hpp"
+
+#define EXPECT_THROW_EXCEPTION(statement, error_msg) \
+  ASSERT_THROW(statement, dds::core::Exception); \
+  try { \
+    statement; \
+  } catch (const dds::core::Exception & err) { \
+    EXPECT_NE(std::string(err.what()).find(error_msg), std::string::npos); \
+  }
+
+namespace
+{
+template<class T>
+void make_sample_(T & sample, const int32_t cnt)
+{
+  (void) sample;
+  (void) cnt;
+  throw std::runtime_error("make_sample_ called on unsupported type");
+}
+
+template<>
+void make_sample_(Space::Type1 & sample, const int32_t cnt)
+{
+  sample.long_1(cnt);
+  sample.long_2(cnt + 1);
+  sample.long_3(cnt + 2);
+}
+
+template<>
+void make_sample_(HelloWorldData::Msg & sample, const int32_t cnt)
+{
+  sample.userID(cnt);
+  sample.message(std::to_string(cnt));
+}
+
+template<>
+void make_sample_(Bounded::Msg & sample, const int32_t cnt)
+{
+  sample.bounded_string(std::to_string(cnt));
+  sample.boolean_sequence().reserve(255);
+  std::fill(sample.boolean_sequence().begin(), sample.boolean_sequence().begin() + 255, true);
+  sample.bounded_sequence().reserve(255);
+  std::fill(sample.bounded_sequence().begin(), sample.bounded_sequence().begin() + 255, cnt);
+}
+}
+/**
+ * Fixture for the shared memory tests with RouDi
+ */
+template<class T>
+class SharedMemoryTest : public RouDi_GTest
+{
+public:
+  using TopicType = T;
+
+  dds::domain::DomainParticipant participant;
+  dds::sub::Subscriber subscriber;
+  dds::pub::Publisher publisher;
+  dds::topic::Topic<T> topic;
+  dds::sub::DataReader<T> reader;
+  dds::pub::DataWriter<T> writer;
+  dds::sub::cond::ReadCondition rc;
+  dds::core::cond::WaitSet waitset;
+
+  SharedMemoryTest()
+  : participant(dds::core::null),
+    subscriber(dds::core::null),
+    publisher(dds::core::null),
+    topic(dds::core::null),
+    reader(dds::core::null),
+    writer(dds::core::null),
+    rc(dds::core::null)
+  {
+  }
+
+  void SetUp()
+  {
+  }
+
+  void CreateParticipant()
+  {
+    if (this->participant == dds::core::null) {
+      // configure cyclone to enable shared memory communication
+      ddsrt_setenv(
+        "CYCLONEDDS_URI",
+        "<CycloneDDS><Domain><SharedMemory><Enable>true</Enable><SubQueueCapacity>256</SubQueueCapacity><SubHistoryRequest>16</SubHistoryRequest><PubHistoryCapacity>16</PubHistoryCapacity><LogLevel>verbose</LogLevel></SharedMemory></Domain></CycloneDDS>");
```

Comment:
Better use the constructor that allows you to provide the configuration as a string.